src/protocols/http/hpr_http_roaming.erl

```diff
@@ -0,0 +1,506 @@
+    NetID :: netid_num(),
+    Uplinks :: list(packet()),
+    TransactionID :: integer(),
+    DedupWindowSize :: non_neg_integer(),
+    Destination :: binary(),
+    FlowType :: sync | async
+) -> prstart_req().
+make_uplink_payload(
+    NetID,
+    Uplinks,
+    TransactionID,
+    DedupWindowSize,
+    Destination,
+    FlowType
+) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = ResponseStream
+    } = select_best(Uplinks),
+    Payload = hpr_packet_up:payload(PacketUp),
+    PacketTime = hpr_packet_up:timestamp(PacketUp),
+
+    Region = hpr_packet_up:region(PacketUp),
+    DataRate = hpr_packet_up:datarate(PacketUp),
+    Frequency = hpr_packet_up:frequency_mhz(PacketUp),
+
+    {RoutingKey, RoutingValue} = routing_key_and_value(PacketUp),
+
+    Token = make_uplink_token(TransactionID, Region, PacketTime, Destination, FlowType),
+
+    ok = hpr_http_roaming_utils:insert_handler(TransactionID, ResponseStream),
+
+    VersionBase = #{
+        'ProtocolVersion' => <<"1.1">>,
+        'SenderNSID' => <<"">>,
+        'DedupWindowSize' => DedupWindowSize
+    },
+
+    VersionBase#{
+        'SenderID' => <<"0xC00053">>,
+        'ReceiverID' => hpr_http_roaming_utils:hexstring(NetID),
+        'TransactionID' => TransactionID,
+        'MessageType' => <<"PRStartReq">>,
+        'PHYPayload' => hpr_http_roaming_utils:binary_to_hexstring(Payload),
+        'ULMetaData' => #{
+            RoutingKey => RoutingValue,
+            'DataRate' => hpr_lorawan:datarate_to_index(Region, DataRate),
+            'ULFreq' => Frequency,
+            'RecvTime' => hpr_http_roaming_utils:format_time(GatewayTime),
+            'RFRegion' => Region,
+            'FNSULToken' => Token,
+            'GWCnt' => erlang:length(Uplinks),
+            'GWInfo' => lists:map(fun gw_info/1, Uplinks)
+        }
+    }.
+
+-spec routing_key_and_value(PacketUp :: hpr_packet_up:packet()) -> {atom(), binary()}.
+routing_key_and_value(PacketUp) ->
+    PacketType = hpr_packet_up:type(PacketUp),
+    {RoutingKey, RoutingValue} =
+        case PacketType of
+            {join_req, {_AppEUI, DevEUI}} ->
+                {'DevEUI', encode_deveui(DevEUI)};
+            {uplink, DevAddr} ->
+                {'DevAddr', encode_devaddr(DevAddr)}
+        end,
+    {RoutingKey, RoutingValue}.
+
+%% ------------------------------------------------------------------
+%% Downlink
+%% ------------------------------------------------------------------
+
+-spec handle_message(prstart_ans() | xmitdata_req()) ->
+    ok
+    | {downlink, xmitdata_ans(), downlink(), {dest_url(), flow_type()}}
+    | {join_accept, downlink()}
+    | {error, any()}.
+handle_message(#{<<"MessageType">> := MT} = M) ->
+    case MT of
+        <<"PRStartAns">> ->
+            handle_prstart_ans(M);
+        <<"XmitDataReq">> ->
+            handle_xmitdata_req(M);
+        _Err ->
+            throw({bad_message, M})
+    end.
```

Comment:
Do we need a catch all if `M` doesn't have a `<<"MessageType">>` key?

Reply:
We want to fail noisily.